src/ecs/src/world.rs

```diff
@@ -21,29 +23,78 @@ impl World {
         World {
             components: Components::new(),
             entities: Entities::new(),
+            ent_data: BTreeMap::new(),
         }
     }
 
     /// Creates a new entity in the world and returns a handle to it.
     pub fn create_entity(&mut self) -> Entity {
         let id = self.entities.create();
-        self.components.insert(id.clone(), AnyVec::new());
+        self.ent_data.insert(id, EntityData::new());
         id
     }
 
-    /// Destroys a given entity and deallocates its components.
+    /// Destroys a given entity and removes its components.
     pub fn destroy_entity(&mut self, entity: Entity) {
-        self.components.remove(&entity);
         self.entities.destroy(entity);
+        if let Some(data) = self.ent_data.remove(&entity) {
+            for (a, b) in data {
+                self.remove_component_type(a, entity);
+            }
+        }
     }
 
     /// Attaches a component to an entity.
-    pub fn insert_component<T: Any>(&mut self, entity: Entity, comp: T) {
-        if self.entities.is_alive(entity) {
-            if let Some(c) = self.components.get_mut(&entity) {
-                c.push(comp);
+    pub fn insert_component<T: Any>(&mut self, entity: Entity, comp: T) -> Option<usize> {
```

Comment:
Just a question, what is the `Option<usize>` return value used for? The variable names in this function don't really give any hints.